### ISSUE TITLE
[$4000] iOS - Sign in  - Text entered inside the email address field is shaking

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#9459]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to App
2. Start to enter any email address in the field

## Expected Result:

_Note: As this issue also exists in the react-native codebase, solutions that fix the problem at the root are preferred_

The text is entered smoothly, without jumps

## Actual Result:
The text are shaking when entering email address

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- iOS

**Version Number:** v1.1.77-2

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/173675127-61e6dcc5-9754-4399-9f9a-13bacae0b78a.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** Applause internal team
**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by adelekennedy** [comment]
Posted:
[Internal](https://www.upwork.com/ab/applicants/1539451122830602240/job-details)
[External](https://www.upwork.com/jobs/~0153a66df54fa2768f)

**Comment by Regaron** [comment]
# Proposal

This is happening due to `defaultValue` in BaseTextInput being controlled. So removing the `defaultValue` solves it.

**Comment by Regaron** [comment]
# Proposal

This is happening due to `defaultValue` being controlled in BaseTextInput. We can remove `defaultValue` prop and add `this.input.setNativeProps({text: this.state.value})` on BaseTextInput `componentDidMount`.

**Comment by parasharrajat** [comment]
@Regaron Would you mind putting up more light on your proposal? I am not able to pinpoint where exactly is the issue. 
1. Where is this `defaultValue`? How is this causing the issue?

**Comment by Regaron** [comment]
> @Regaron Would you mind putting up more light on your proposal? I am not able to pinpoint where exactly is the issue.
> 
> 1. Where is this `defaultValue`? How is this causing the issue?

https://github.com/Expensify/App/blob/b38693e78e955ff3c19ebe5681390f85d4c5b2d4/src/components/TextInput/BaseTextInput.js#L255

**Comment by Regaron** [comment]
> @Regaron Would you mind putting up more light on your proposal? I am not able to pinpoint where exactly is the issue.
> 
> 1. Where is this `defaultValue`? How is this causing the issue?

https://github.com/Expensify/App/blob/b38693e78e955ff3c19ebe5681390f85d4c5b2d4/src/components/TextInput/BaseTextInput.js#L255

This line is making the input controlled. And controlled input seems to have the above mentioned problem.

**Comment by mdneyazahmad** [comment]
## Proposal

To understand the issue better I created a bare minimum expo project and this issue also exists there. Android works fine but iOS has the issue. I think this is a bug in react native itself as the two platforms behave differently.

When we have text larger than the space available and explicitly set the curser in between and typing the cursor position does not change and it prepends the text below the cursor but the expected behavior should add the text after the cursor till the end of the text input.

`defaultValue` is used to provide an initial value and it should not be updated to the current input text. My proposal is to accept this prop from `TextInput` component.

https://github.com/Expensify/App/blob/b38693e78e955ff3c19ebe5681390f85d4c5b2d4/src/components/TextInput/BaseTextInput.js#L255

```diff
- defaultValue={this.state.value} 
+ defaultValue={this.props..defaultValue} 
```

https://user-images.githubusercontent.com/77761491/175280895-b2a52640-6176-48ba-9146-b95d9ccc2242.mp4

https://user-images.githubusercontent.com/77761491/175280007-2f22b1da-03da-4713-aa18-b20ea52fe1bd.mp4


<details><summary>Expo reproduction code</summary>

```javascript
import React from "react";
import { SafeAreaView, StyleSheet, TextInput } from "react-native";

const UselessTextInput = () => {
  const [text, onChangeText] = React.useState('Initial Value');
  
  return (
    <SafeAreaView style={styles.container}>
      <TextInput
        style={styles.input}
        defaultValue={text}
        onChangeText={onChangeText}
        value={text}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
  },
  input: {
    height: 40,
    margin: 12,
    borderWidth: 1,
    padding: 10,
  },
});

export default UselessTextInput;
```
</details>

**Comment by parasharrajat** [comment]
Thanks, I will do some research and get back to you.

**Comment by adelekennedy** [comment]
still pending research - no action for now!

**Comment by adelekennedy** [comment]
@parasharrajat should we put this on hold for now?

**Comment by parasharrajat** [comment]
I didn't find much detail about the effect of updating the `defaultValue`. This should not have any effect as per the docs. 

Unfortunately, I can't accept any proposal until they explain the reason why it is causing the issue only on iOS. Meanwhile, I am also looking into the same matter to understand the reason.

**Comment by adelekennedy** [comment]
Still researching - not overdue

**Comment by adelekennedy** [comment]
any updates here @parasharrajat?

**Comment by parasharrajat** [comment]
Waiting on responses to explain the reason behind the issue or new proposals.

**Comment by mdneyazahmad** [comment]
@parasharrajat In general default value should not be updated or if so, it should not have any effect. As I have pointed out this is an issue in `react-native` itself. I do not see an issue with the same that may be because no one updates it with default value prop. Do you want us to find the reason in `react-native` codebase? Let me know if I am missing something.

**Comment by parasharrajat** [comment]
Thanks for bringing that up. 

> Do you want us to find the reason in react-native codebase? 

Kind of. It would be good to see some research on this before we move ahead and implement it. 
- Any existing issues? 
- Any discussions on the same?
- A minimal reproduction of the bug in RN to confirm if this really a RN bug. 
- You can also show some recommended material which shows that we should not update defaultvalue prop to prove that this is our mistake.

ETC. 

If this really a issue with RN then we should fix it there to help the cummunity. 

I tried to find something on the same but had no luck.

**Comment by adelekennedy** [comment]
still researching this - a few different proposals are above

**Comment by mdneyazahmad** [comment]
@parasharrajat I debug this and found an issue in `react-native` https://github.com/facebook/react-native/issues/34276

This issue happens when you make controlled text input either `value or defaultValue` produces this issue.

```
<TextInput value={text} onChangeText={value => setText(value)} />

OR

<TextInput defaultValue={text} onChangeText={value => setText(value)} />

```

If we want to fix this issue we have to remove these props and make uncontrolled text input or it is fixed in `react-native` itself.

**Comment by adelekennedy** [comment]
it looks like we're still reviewing proposals

**Comment by NikkiWines** [comment]
Hmmm, so actually a sort of similar bug exists on mWeb (iOS) where the screen shakes to keep the cursor in view - the iOS app one is effectively doing the same thing just in smaller increments (every letter instead of once at the end) but functionally the "bug" exists on both. Which kind of makes me think this isn't really that much of a bug at all. What do you think?

https://user-images.githubusercontent.com/16747822/184838474-624ece49-52f0-41be-9055-5e67cd6cd63e.MP4

Given that I lean towards option 3. But if we do want to fix it then option 2 seems fine to me

**Comment by NikkiWines** [comment]
I agree they're not identical issues - just noting that the behavior (moving the cursor to stay in view for long text inputs) exists on both platforms, albeit in slightly different forms. 

Fixing it in the app seems like the best way to go for now 👍

**Comment by mdneyazahmad** [comment]
@parasharrajat As long as we do not update the defaultValue that should work and I believe we won't be updating that way also react-native fabric architecture does not support setNativeProps (post fabric).

```
// This will create problem
<TextInputParent defaultValue={value} onChange={onChangeValue} /> 

// This will not create any issue
<TextInputParent defaultValue="Default Value" onChange={onChangeValue} />
```

**Comment by NikkiWines** [comment]
+1 agreed with your assessment, @Regaron's proposal seems fine. @adelekennedy would you mind hiring them for this role? Thank you!

**Comment by adelekennedy** [comment]
@Regaron will you apply [here](https://www.upwork.com/jobs/~01603d4b51b32e9b7a)?

**Comment by Regaron** [comment]
Applied in upwork.

**Comment by adelekennedy** [comment]
@parasharrajat I forget do you prefer to apply for the role or can I send you an offer?

**Comment by Regaron** [comment]
#10147 Changed all Input to be controlled.

**Comment by adelekennedy** [comment]
@parasharrajat @Regaron will you reapply [here](https://www.upwork.com/jobs/~01a6e38c08aa3d14cb)? the post expired.

**Comment by Regaron** [comment]
I don't have any other idea to fix this currently.

**Comment by NikkiWines** [comment]
@parasharrajat were you able to reproduce this issue?

**Comment by NikkiWines** [comment]
Ok cool, @adelekennedy can we re-open this role on UpWork? Thank you

**Comment by NikkiWines** [comment]
Waiting on proposals

**Comment by NikkiWines** [comment]
Still waiting on proposals

**Comment by adelekennedy** [comment]
Not overdue - waiting on proposals. We've capped this at 2k

**Comment by NikkiWines** [comment]
^^ see above message

**Comment by arielgreen** [comment]
@adelekennedy can you provide an assessment on what needs to happen to move this forward? Is there a reason we've capped the price or should we push forward and increase again?

**Comment by adelekennedy** [comment]
we capped based on a prior 1:1 convo between Nikki and I, let's move forward with increasing this weekly

**Comment by adelekennedy** [comment]
@abekkala I'm mostly ooo next week - reassigning this!

**Comment by abekkala** [comment]
Hmm - just assigned. Not overdue MELVIN!

**Comment by NikkiWines** [comment]
I'm not sure how we determine when to double or not. 

In my opinion, given this is a fairly edge-case scenario and more of a minor inconvenience than an app-breaking bug, I think $4000 is an appropriate price.

**Comment by abekkala** [comment]
Ok. I'll keep it at Adele's last updated price and hope for some proposals!

**Comment by sakulz** [comment]
**Proposal**

Problem is due to controlled component, as the value gets updated time to time by controlled way, due to that shaking appears on the input.

Solution: Make BaseTextInput component uncontrolled, and add a method to forceUpdate value from parent; if required in certain cases.

Here below is the screen recording to show the change in action.
When pressing continue button -> Parent forceUpdate child with value = "";




https://user-images.githubusercontent.com/118268491/201931599-935f0505-9208-438b-a32b-62b9c42247c9.mov

**Comment by parasharrajat** [comment]
@sakulz We are moving to Fabric so do you think you proposal is still valid? 

Also, a solution on RN codebase is preferred as this issue exists with RN.

--- 

@NikkiWines Can you please update the issue Expected Result section to mention that we RN fix is preferred? Thanks.

**Comment by abekkala** [comment]
@kevinksullivan I did ask about this above. It was decided by @NikkiWines to hold at $4000.

**Comment by abekkala** [comment]
@kevinksullivan I did ask about this above. https://github.com/Expensify/App/issues/9459#issuecomment-1313677807
It was decided by @NikkiWines to hold at $4000.

**Comment by JmillsExpensify** [comment]
@NikkiWines To help clear out the extensive backlog of /App bugs, we're putting the spotlight on all bugs older than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:

 - Decide whether any proposals currently meet our guidelines and can be approved as-is
 - For any that can't, please take this issue internal and treat it as one of your highest priorities
 - If you have any questions, don't hesitate to start a discussion in #bug-zero

Thanks everyone!

**Comment by NikkiWines** [comment]
This issue was brought up in #bug-zero [today](https://expensify.slack.com/archives/C01SKUP7QR0/p1668689378986189) and I believe the consensus is that this isn't really a true bug. I said effectively as much [here](https://github.com/Expensify/App/issues/9459#issuecomment-1216335157). 

Curious to your thoughts here @JmillsExpensify.

**Comment by JmillsExpensify** [comment]
Thanks @NikkiWines. I agree with the discussion in Slack. There are reproduction steps, however, one has to make a mistake on a very long email address, and then go back to fix it. More to the point, this is better fixed upstream in RN, and enabling Fabric may remove this issue entirely. So I'm good with closing it as a result.

**Comment by michaelhaxhiu** [comment]
@abekkala we are going to compensate Rajat for reviewing several proposals in this issue (before it was closed).

More context [here in Slack ](https://expensify.slack.com/archives/C02NK2DQWUX/p1670540622837929?thread_ts=1668807435.007649&cid=C02NK2DQWUX)

I'll assign it to help see it through.

**Comment by michaelhaxhiu** [comment]
https://www.upwork.com/jobs/~01245bd6daa045ce46

**Comment by trjExpensify** [comment]
@michaelhaxhiu I'm removing the `Help wanted` label to avoid confusion here.

**Comment by michaelhaxhiu** [comment]
woopsie, thanks TRJ

**Comment by michaelhaxhiu** [comment]
Paid.
